cpp/src/arrow/util/async_generator.h

```diff
@@ -336,6 +345,68 @@ class ReadaheadGenerator {
   std::queue<Future<T>> readahead_queue_;
 };
 
+template <typename T>
+class PushGenerator {
+  struct State {
+    util::Mutex mutex;
+    std::deque<Result<T>> result_q;
+    util::optional<Future<T>> consumer_fut;
+    bool finished = false;
+  };
+
+  struct Generator {
+    const std::shared_ptr<State> state_;
+
+    Future<T> operator()() {
+      auto lock = state_->mutex.Lock();
+      assert(!state_->consumer_fut.has_value());  // Non-reentrant
+      if (!state_->result_q.empty()) {
+        auto fut = Future<T>::MakeFinished(std::move(state_->result_q.front()));
+        state_->result_q.pop_front();
+        return fut;
+      }
+      if (state_->finished) {
+        return AsyncGeneratorEnd<T>();
+      }
+      auto fut = Future<T>::Make();
+      state_->consumer_fut = fut;
+      return fut;
+    }
+  };
+
+ public:
+  PushGenerator() : state_(std::make_shared<State>()) {}
+
+  void Push(Result<T> result) {
```

Comment:
You could maybe check if `result` is not ok and mark finished to true (potentially even clearing out the result q) and then on future pushes simply return immediately if finished is true.  I can see where you question on Zulip came from now.  The only disadvantage I can see to this approach is potentially wasted memory keeping blocks around that are invalid.

Reply:
Indeed, I could do that. The underlying question is: should an error always terminate an async generator? It doesn't seem that obvious to me.